File: LeetCodeTasks/39_combination-sum/solutions/cpp/solution.cpp
#include <vector>

class Solution {
public:
    void backtrack(std::vector<int>& candidates, int target, std::vector<std::vector<int>>& result, std::vector<int>& current, int index) {
        if (target == 0) {
            result.push_back(current);
            return;
        }
        
        for (int i = index; i < candidates.size(); i++) {
            if (candidates[i] > target) continue;
            current.push_back(candidates[i]);
            backtrack(candidates, target - candidates[i], result, current, i);
            current.pop_back();
        }
    }
    
    std::vector<std::vector<int>> combinationSum(std::vector<int>& candidates, int target) {
        std::vector<std::vector<int>> result;
        std::vector<int> current;
        backtrack(candidates, target, result, current, 0);
        return result;
    }
};

File: LeetCodeTasks/39_combination-sum/solutions/java/solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> results = new ArrayList<>();
        backtrack(candidates, target, 0, new ArrayList<>(), results);
        return results;
    }
    
    private void backtrack(int[] candidates, int remaining, int index, List<Integer> current, List<List<Integer>> results) {
        if (remaining == 0) {
            results.add(new ArrayList<>(current));
            return;
        }
        
        for (int i = index; i < candidates.length; i++) {
            if (candidates[i] <= remaining) {
                current.add(candidates[i]);
                backtrack(candidates, remaining - candidates[i], i, current, results);
                current.remove(current.size() - 1);
            }
        }
    }
}
